src/rgw/rgw_data_sync.cc

```diff
@@ -5211,12 +5310,18 @@ int RGWSyncBucketCR::operate()
           }
         }
 
-        yield call(new RGWSyncBucketShardCR(sc, data_lease_cr, sync_pair,
-                                            sync_pipe, bucket_status.state,
-                                            tn, progress));
-        if (retcode < 0) {
-          return set_cr_error(retcode);
+        if (bucket_status.shards_done_with_gen.at(sync_pair.source_bs.shard_id) != true) {
```

Comment:
its probably worth checking that `shard_id < shards_done_with_gen.size()` so we don't crash here

and let's `return set_cr_done()` immediately in this case, instead of relying on the loop to end and return done later; this logic in `RGWSyncBucketCR` is already complicated, and i expect more changes here to support `bucket sync disable`

```c++
if (sync_pair.source_bs.shard_id >= bucket_status.shards_done_with_gen.size()) {
  tn->log(1, SSTR("bucket shard " << sync_pair.source_bs << " index out of bounds"));
  return set_cr_done(); // return success so we don't retry
}
if (bucket_status.shards_done_with_gen[sync_pair.source_bs.shard_id]) {
  tn->log(10, SSTR("bucket shard " << sync_pair.source_bs << " of gen " <<
                   gen << " already synced."));
  return set_cr_done();
}
yield call(new RGWSyncBucketShardCR(...
```

Reply:
thanks, done